File: snippets/design-patterns/factory/FactoryPattern.cpp
/**
 * Factory Pattern Implementation in C++
 *
 * The Factory Pattern is a creational design pattern that provides an interface for creating
 * objects in a superclass, but allows subclasses to alter the type of objects that will be created.
 *
 * This example demonstrates a Vehicle Factory that can create different types of vehicles
 * (Car, Motorcycle, Truck) based on the client's requirements.
 */

#include <iostream>
#include <memory>
#include <string>
#include <vector>

// Abstract Product - Vehicle
class Vehicle {
public:
    Vehicle(const std::string& make, const std::string& model, int year)
        : make_(make), model_(model), year_(year) {}

    virtual ~Vehicle() = default;

    virtual std::string getInfo() const {
        return std::to_string(year_) + " " + make_ + " " + model_;
    }

    virtual std::string start() const {
        return getInfo() + " is starting...";
    }

    virtual std::string stop() const {
        return getInfo() + " is stopping...";
    }

protected:
    std::string make_;
    std::string model_;
    int year_;
};

// Concrete Products
class Car : public Vehicle {
public:
    Car(const std::string& make, const std::string& model, int year, int doors = 4)
        : Vehicle(make, model, year), doors_(doors) {}

    std::string getInfo() const override {
        return Vehicle::getInfo() + " (" + std::to_string(doors_) + "-door car)";
    }

    std::string drive() const {
        return getInfo() + " is driving on the road.";
    }

private:
    int doors_;
};

class Motorcycle : public Vehicle {
public:
    Motorcycle(const std::string& make, const std::string& model, int year, int engineSize)
        : Vehicle(make, model, year), engineSize_(engineSize) {}

    std::string getInfo() const override {
        return Vehicle::getInfo() + " (" + std::to_string(engineSize_) + "cc motorcycle)";
    }

    std::string ride() const {
        return getInfo() + " is riding at high speed.";
    }

private:
    int engineSize_;
};

class Truck : public Vehicle {
public:
    Truck(const std::string& make, const std::string& model, int year, double capacity)
        : Vehicle(make, model, year), capacity_(capacity) {}

    std::string getInfo() const override {
        return Vehicle::getInfo() + " (" + std::to_string(capacity_) + " ton truck)";
    }

    std::string haul() const {
        return getInfo() + " is hauling cargo.";
    }

private:
    double capacity_;
};

// Simple Factory
class VehicleFactory {
public:
    enum VehicleType {
        CAR,
        MOTORCYCLE,
        TRUCK
    };

    static std::unique_ptr<Vehicle> createVehicle(VehicleType type,
                                                  const std::string& make,
                                                  const std::string& model,
                                                  int year,
                                                  const std::vector<double>& options = {}) {
        switch (type) {
            case CAR:
                return std::make_unique<Car>(make, model, year,
                                            options.size() > 0 ? static_cast<int>(options[0]) : 4);
            case MOTORCYCLE:
                return std::make_unique<Motorcycle>(make, model, year,
                                                  options.size() > 0 ? static_cast<int>(options[0]) : 250);
            case TRUCK:
                return std::make_unique<Truck>(make, model, year,
                                              options.size() > 0 ? options[0] : 5.0);
            default:
                throw std::invalid_argument("Unknown vehicle type");
        }
    }
};

// Factory Method Pattern Implementation
class VehicleFactoryMethod {
public:
    virtual ~VehicleFactoryMethod() = default;

    virtual std::unique_ptr<Vehicle> createVehicle(const std::string& make,
                                                 const std::string& model,
                                                 int year,
                                                 const std::vector<double>& options = {}) = 0;

    std::unique_ptr<Vehicle> registerVehicle(const std::string& make,
                                           const std::string& model,
                                           int year,
                                           const std::vector<double>& options = {}) {
        // Common operations for all vehicles
        auto vehicle = createVehicle(make, model, year, options);
        std::cout << "Registering " << vehicle->getInfo() << std::endl;
        std::cout << "Assigning license plate" << std::endl;
        return vehicle;
    }
};

// Concrete Factories
class CarFactory : public VehicleFactoryMethod {
public:
    std::unique_ptr<Vehicle> createVehicle(const std::string& make,
                                         const std::string& model,
                                         int year,
                                         const std::vector<double>& options = {}) override {
        return std::make_unique<Car>(make, model, year,
                                    options.size() > 0 ? static_cast<int>(options[0]) : 4);
    }
};

class MotorcycleFactory : public VehicleFactoryMethod {
public:
    std::unique_ptr<Vehicle> createVehicle(const std::string& make,
                                         const std::string& model,
                                         int year,
                                         const std::vector<double>& options = {}) override {
        return std::make_unique<Motorcycle>(make, model, year,
                                          options.size() > 0 ? static_cast<int>(options[0]) : 250);
    }
};

class TruckFactory : public VehicleFactoryMethod {
public:
    std::unique_ptr<Vehicle> createVehicle(const std::string& make,
                                         const std::string& model,
                                         int year,
                                         const std::vector<double>& options = {}) override {
        return std::make_unique<Truck>(make, model, year,
                                      options.size() > 0 ? options[0] : 5.0);
    }
};

// Abstract Factory Pattern Implementation
// Parts interfaces
class Engine {
public:
    Engine(const std::string& type, int horsepower)
        : type_(type), horsepower_(horsepower) {}

    virtual ~Engine() = default;

    std::string getSpecs() const {
        return type_ + " engine with " + std::to_string(horsepower_) + "hp";
    }

private:
    std::string type_;
    int horsepower_;
};

class Transmission {
public:
    Transmission(const std::string& type, int gears)
        : type_(type), gears_(gears) {}

    virtual ~Transmission() = default;

    std::string getSpecs() const {
        return type_ + " transmission with " + std::to_string(gears_) + " gears";
    }

private:
    std::string type_;
    int gears_;
};

class Chassis {
public:
    Chassis(const std::string& material, double weight)
        : material_(material), weight_(weight) {}

    virtual ~Chassis() = default;

    std::string getSpecs() const {
        return material_ + " chassis weighing " + std::to_string(weight_) + "kg";
    }

private:
    std::string material_;
    double weight_;
};

// Abstract Factory
class VehiclePartsFactory {
public:
    virtual ~VehiclePartsFactory() = default;
    virtual std::unique_ptr<Engine> createEngine() = 0;
    virtual std::unique_ptr<Transmission> createTransmission() = 0;
    virtual std::unique_ptr<Chassis> createChassis() = 0;
};

// Concrete Abstract Factories
class SportVehiclePartsFactory : public VehiclePartsFactory {
public:
    std::unique_ptr<Engine> createEngine() override {
        return std::make_unique<Engine>("V8", 450);
    }

    std::unique_ptr<Transmission> createTransmission() override {
        return std::make_unique<Transmission>("Manual", 6);
    }

    std::unique_ptr<Chassis> createChassis() override {
        return std::make_unique<Chassis>("Carbon Fiber", 120);
    }
};

class EconomyVehiclePartsFactory : public VehiclePartsFactory {
public:
    std::unique_ptr<Engine> createEngine() override {
        return std::make_unique<Engine>("Inline-4", 180);
    }

    std::unique_ptr<Transmission> createTransmission() override {
        return std::make_unique<Transmission>("Automatic", 5);
    }

    std::unique_ptr<Chassis> createChassis() override {
        return std::make_unique<Chassis>("Steel", 300);
    }
};

class HeavyDutyVehiclePartsFactory : public VehiclePartsFactory {
public:
    std::unique_ptr<Engine> createEngine() override {
        return std::make_unique<Engine>("Diesel V6", 350);
    }

    std::unique_ptr<Transmission> createTransmission() override {
        return std::make_unique<Transmission>("Manual", 8);
    }

    std::unique_ptr<Chassis> createChassis() override {
        return std::make_unique<Chassis>("Reinforced Steel", 800);
    }
};

// Vehicle Assembler - Uses the Abstract Factory
class VehicleAssembler {
public:
    VehicleAssembler(std::unique_ptr<VehiclePartsFactory> factory)
        : partsFactory_(std::move(factory)) {}

    void assembleVehicle() {
        auto engine = partsFactory_->createEngine();
        auto transmission = partsFactory_->createTransmission();
        auto chassis = partsFactory_->createChassis();

        std::cout << "Assembling vehicle with:" << std::endl;
        std::cout << "- " << engine->getSpecs() << std::endl;
        std::cout << "- " << transmission->getSpecs() << std::endl;
        std::cout << "- " << chassis->getSpecs() << std::endl;
    }

private:
    std::unique_ptr<VehiclePartsFactory> partsFactory_;
};

// Client code demonstration
void clientCode() {
    std::cout << "===== Simple Factory Pattern =====" << std::endl;

    auto car = VehicleFactory::createVehicle(VehicleFactory::CAR, "Toyota", "Camry", 2023, {4});
    auto motorcycle = VehicleFactory::createVehicle(VehicleFactory::MOTORCYCLE, "Honda", "CBR", 2023, {600});
    auto truck = VehicleFactory::createVehicle(VehicleFactory::TRUCK, "Ford", "F-150", 2023, {3.0});

    std::cout << car->getInfo() << std::endl;
    std::cout << static_cast<Car*>(car.get())->drive() << std::endl;

    std::cout << motorcycle->getInfo() << std::endl;
    std::cout << static_cast<Motorcycle*>(motorcycle.get())->ride() << std::endl;

    std::cout << truck->getInfo() << std::endl;
    std::cout << static_cast<Truck*>(truck.get())->haul() << std::endl;

    std::cout << "\n===== Factory Method Pattern =====" << std::endl;

    CarFactory carFactory;
    MotorcycleFactory motorcycleFactory;
    TruckFactory truckFactory;

    auto newCar = carFactory.registerVehicle("BMW", "3 Series", 2023, {2});
    auto newMotorcycle = motorcycleFactory.registerVehicle("Ducati", "Monster", 2023, {821});
    auto newTruck = truckFactory.registerVehicle("Volvo", "VNL", 2023, {20.0});

    std::cout << static_cast<Car*>(newCar.get())->drive() << std::endl;
    std::cout << static_cast<Motorcycle*>(newMotorcycle.get())->ride() << std::endl;
    std::cout << static_cast<Truck*>(newTruck.get())->haul() << std::endl;

    std::cout << "\n===== Abstract Factory Pattern =====" << std::endl;

    std::cout << "Building a sports car:" << std::endl;
    VehicleAssembler sportCarAssembler(std::make_unique<SportVehiclePartsFactory>());
    sportCarAssembler.assembleVehicle();

    std::cout << "\nBuilding an economy car:" << std::endl;
    VehicleAssembler economyCarAssembler(std::make_unique<EconomyVehiclePartsFactory>());
    economyCarAssembler.assembleVehicle();

    std::cout << "\nBuilding a heavy duty truck:" << std::endl;
    VehicleAssembler heavyDutyTruckAssembler(std::make_unique<HeavyDutyVehiclePartsFactory>());
    heavyDutyTruckAssembler.assembleVehicle();
}

// Main function, run the example
int main() {
    try {
        clientCode();
    } catch (const std::exception& e) {
        std::cerr << "Exception: " << e.what() << std::endl;
        return 1;
    }
    return 0;
}

File: snippets/design-patterns/factory/factory_pattern.rs
/**
 * Factory Pattern Implementation in Rust
 *
 * The Factory Pattern is a creational design pattern that provides an interface for creating
 * objects in a superclass, but allows subclasses to alter the type of objects that will be created.
 *
 * This example demonstrates a Vehicle Factory that can create different types of vehicles
 * (Car, Motorcycle, Truck) based on the client's requirements.
 */

use std::fmt;

// Abstract Product - Vehicle Trait
trait Vehicle {
    fn get_info(&self) -> String;
    fn start(&self) -> String {
        format!("{} is starting...", self.get_info())
    }
    fn stop(&self) -> String {
        format!("{} is stopping...", self.get_info())
    }
}

// Concrete Products
struct Car {
    make: String,
    model: String,
    year: u32,
    doors: u32,
}

impl Car {
    fn new(make: &str, model: &str, year: u32, doors: u32) -> Self {
        Car {
            make: make.to_string(),
            model: model.to_string(),
            year,
            doors,
        }
    }

    fn drive(&self) -> String {
        format!("{} is driving on the road.", self.get_info())
    }
}

impl Vehicle for Car {
    fn get_info(&self) -> String {
        format!("{} {} {} ({}-door car)", self.year, self.make, self.model, self.doors)
    }
}

struct Motorcycle {
    make: String,
    model: String,
    year: u32,
    engine_size: u32,
}

impl Motorcycle {
    fn new(make: &str, model: &str, year: u32, engine_size: u32) -> Self {
        Motorcycle {
            make: make.to_string(),
            model: model.to_string(),
            year,
            engine_size,
        }
    }

    fn ride(&self) -> String {
        format!("{} is riding at high speed.", self.get_info())
    }
}

impl Vehicle for Motorcycle {
    fn get_info(&self) -> String {
        format!(
            "{} {} {} ({}cc motorcycle)",
            self.year, self.make, self.model, self.engine_size
        )
    }
}

struct Truck {
    make: String,
    model: String,
    year: u32,
    capacity: f64,
}

impl Truck {
    fn new(make: &str, model: &str, year: u32, capacity: f64) -> Self {
        Truck {
            make: make.to_string(),
            model: model.to_string(),
            year,
            capacity,
        }
    }

    fn haul(&self) -> String {
        format!("{} is hauling cargo.", self.get_info())
    }
}

impl Vehicle for Truck {
    fn get_info(&self) -> String {
        format!(
            "{} {} {} ({} ton truck)",
            self.year, self.make, self.model, self.capacity
        )
    }
}

// Simple Factory
enum VehicleType {
    Car,
    Motorcycle,
    Truck,
}

struct VehicleFactory;

impl VehicleFactory {
    fn create_car(make: &str, model: &str, year: u32, doors: u32) -> Box<dyn Vehicle> {
        Box::new(Car::new(make, model, year, doors))
    }

    fn create_motorcycle(make: &str, model: &str, year: u32, engine_size: u32) -> Box<dyn Vehicle> {
        Box::new(Motorcycle::new(make, model, year, engine_size))
    }

    fn create_truck(make: &str, model: &str, year: u32, capacity: f64) -> Box<dyn Vehicle> {
        Box::new(Truck::new(make, model, year, capacity))
    }

    fn create_vehicle(
        vehicle_type: VehicleType,
        make: &str,
        model: &str,
        year: u32,
        options: &[f64],
    ) -> Box<dyn Vehicle> {
        match vehicle_type {
            VehicleType::Car => {
                let doors = if options.is_empty() { 4 } else { options[0] as u32 };
                Self::create_car(make, model, year, doors)
            }
            VehicleType::Motorcycle => {
                let engine_size = if options.is_empty() { 250 } else { options[0] as u32 };
                Self::create_motorcycle(make, model, year, engine_size)
            }
            VehicleType::Truck => {
                let capacity = if options.is_empty() { 5.0 } else { options[0] };
                Self::create_truck(make, model, year, capacity)
            }
        }
    }
}

// Factory Method Pattern Implementation
trait VehicleFactoryMethod {
    fn create_vehicle(&self, make: &str, model: &str, year: u32, options: &[f64]) -> Box<dyn Vehicle>;

    fn register_vehicle(&self, make: &str, model: &str, year: u32, options: &[f64]) -> Box<dyn Vehicle> {
        // Common operations for all vehicles
        let vehicle = self.create_vehicle(make, model, year, options);
        println!("Registering {}", vehicle.get_info());
        println!("Assigning license plate");
        vehicle
    }
}

// Concrete Factories
struct CarFactory;

impl VehicleFactoryMethod for CarFactory {
    fn create_vehicle(&self, make: &str, model: &str, year: u32, options: &[f64]) -> Box<dyn Vehicle> {
        let doors = if options.is_empty() { 4 } else { options[0] as u32 };
        Box::new(Car::new(make, model, year, doors))
    }
}

struct MotorcycleFactory;

impl VehicleFactoryMethod for MotorcycleFactory {
    fn create_vehicle(&self, make: &str, model: &str, year: u32, options: &[f64]) -> Box<dyn Vehicle> {
        let engine_size = if options.is_empty() { 250 } else { options[0] as u32 };
        Box::new(Motorcycle::new(make, model, year, engine_size))
    }
}

struct TruckFactory;

impl VehicleFactoryMethod for TruckFactory {
    fn create_vehicle(&self, make: &str, model: &str, year: u32, options: &[f64]) -> Box<dyn Vehicle> {
        let capacity = if options.is_empty() { 5.0 } else { options[0] };
        Box::new(Truck::new(make, model, year, capacity))
    }
}

// Abstract Factory Pattern Implementation
// Parts
struct Engine {
    engine_type: String,
    horsepower: u32,
}

impl Engine {
    fn new(engine_type: &str, horsepower: u32) -> Self {
        Engine {
            engine_type: engine_type.to_string(),
            horsepower,
        }
    }

    fn get_specs(&self) -> String {
        format!("{} engine with {}hp", self.engine_type, self.horsepower)
    }
}

struct Transmission {
    transmission_type: String,
    gears: u32,
}

impl Transmission {
    fn new(transmission_type: &str, gears: u32) -> Self {
        Transmission {
            transmission_type: transmission_type.to_string(),
            gears,
        }
    }

    fn get_specs(&self) -> String {
        format!(
            "{} transmission with {} gears",
            self.transmission_type, self.gears
        )
    }
}

struct Chassis {
    material: String,
    weight: f64,
}

impl Chassis {
    fn new(material: &str, weight: f64) -> Self {
        Chassis {
            material: material.to_string(),
            weight,
        }
    }

    fn get_specs(&self) -> String {
        format!("{} chassis weighing {}kg", self.material, self.weight)
    }
}

// Abstract Factory
trait VehiclePartsFactory {
    fn create_engine(&self) -> Engine;
    fn create_transmission(&self) -> Transmission;
    fn create_chassis(&self) -> Chassis;
}

// Concrete Abstract Factories
struct SportVehiclePartsFactory;

impl VehiclePartsFactory for SportVehiclePartsFactory {
    fn create_engine(&self) -> Engine {
        Engine::new("V8", 450)
    }

    fn create_transmission(&self) -> Transmission {
        Transmission::new("Manual", 6)
    }

    fn create_chassis(&self) -> Chassis {
        Chassis::new("Carbon Fiber", 120.0)
    }
}

struct EconomyVehiclePartsFactory;

impl VehiclePartsFactory for EconomyVehiclePartsFactory {
    fn create_engine(&self) -> Engine {
        Engine::new("Inline-4", 180)
    }

    fn create_transmission(&self) -> Transmission {
        Transmission::new("Automatic", 5)
    }

    fn create_chassis(&self) -> Chassis {
        Chassis::new("Steel", 300.0)
    }
}

struct HeavyDutyVehiclePartsFactory;

impl VehiclePartsFactory for HeavyDutyVehiclePartsFactory {
    fn create_engine(&self) -> Engine {
        Engine::new("Diesel V6", 350)
    }

    fn create_transmission(&self) -> Transmission {
        Transmission::new("Manual", 8)
    }

    fn create_chassis(&self) -> Chassis {
        Chassis::new("Reinforced Steel", 800.0)
    }
}

// Vehicle Assembler - Uses the Abstract Factory
struct VehicleAssembler<T: VehiclePartsFactory> {
    parts_factory: T,
}

impl<T: VehiclePartsFactory> VehicleAssembler<T> {
    fn new(parts_factory: T) -> Self {
        VehicleAssembler { parts_factory }
    }

    fn assemble_vehicle(&self) {
        let engine = self.parts_factory.create_engine();
        let transmission = self.parts_factory.create_transmission();
        let chassis = self.parts_factory.create_chassis();

        println!("Assembling vehicle with:");
        println!("- {}", engine.get_specs());
        println!("- {}", transmission.get_specs());
        println!("- {}", chassis.get_specs());
    }
}

// Client code demonstration
fn client_code() {
    println!("===== Simple Factory Pattern =====");

    let car = VehicleFactory::create_vehicle(
        VehicleType::Car,
        "Toyota",
        "Camry",
        2023,
        &[4.0],
    );
    let motorcycle = VehicleFactory::create_vehicle(
        VehicleType::Motorcycle,
        "Honda",
        "CBR",
        2023,
        &[600.0],
    );
    let truck = VehicleFactory::create_vehicle(
        VehicleType::Truck,
        "Ford",
        "F-150",
        2023,
        &[3.0],
    );

    println!("{}", car.get_info());
    // We need to downcast to call specific methods
    if let Some(car) = car.as_any().downcast_ref::<Car>() {
        println!("{}", car.drive());
    }

    println!("{}", motorcycle.get_info());
    if let Some(motorcycle) = motorcycle.as_any().downcast_ref::<Motorcycle>() {
        println!("{}", motorcycle.ride());
    }

    println!("{}", truck.get_info());
    if let Some(truck) = truck.as_any().downcast_ref::<Truck>() {
        println!("{}", truck.haul());
    }

    println!("\n===== Factory Method Pattern =====");

    let car_factory = CarFactory;
    let motorcycle_factory = MotorcycleFactory;
    let truck_factory = TruckFactory;

    let new_car = car_factory.register_vehicle("BMW", "3 Series", 2023, &[2.0]);
    let new_motorcycle = motorcycle_factory.register_vehicle("Ducati", "Monster", 2023, &[821.0]);
    let new_truck = truck_factory.register_vehicle("Volvo", "VNL", 2023, &[20.0]);

    if let Some(car) = new_car.as_any().downcast_ref::<Car>() {
        println!("{}", car.drive());
    }
    if let Some(motorcycle) = new_motorcycle.as_any().downcast_ref::<Motorcycle>() {
        println!("{}", motorcycle.ride());
    }
    if let Some(truck) = new_truck.as_any().downcast_ref::<Truck>() {
        println!("{}", truck.haul());
    }

    println!("\n===== Abstract Factory Pattern =====");

    println!("Building a sports car:");
    let sport_car_assembler = VehicleAssembler::new(SportVehiclePartsFactory);
    sport_car_assembler.assemble_vehicle();

    println!("\nBuilding an economy car:");
    let economy_car_assembler = VehicleAssembler::new(EconomyVehiclePartsFactory);
    economy_car_assembler.assemble_vehicle();

    println!("\nBuilding a heavy duty truck:");
    let heavy_duty_truck_assembler = VehicleAssembler::new(HeavyDutyVehiclePartsFactory);
    heavy_duty_truck_assembler.assemble_vehicle();
}

// Extension trait to allow downcasting
trait AsAny {
    fn as_any(&self) -> &dyn std::any::Any;
}

impl<T: 'static> AsAny for T {
    fn as_any(&self) -> &dyn std::any::Any {
        self
    }
}

// Extend Vehicle trait to include AsAny
trait VehicleExt: Vehicle + AsAny {}

// Implement VehicleExt for all types that implement Vehicle
impl<T: Vehicle + AsAny> VehicleExt for T {}

// Update the Vehicle trait to include AsAny functionality
trait Vehicle: AsAny {
    fn get_info(&self) -> String;
    fn start(&self) -> String {
        format!("{} is starting...", self.get_info())
    }
    fn stop(&self) -> String {
        format!("{} is stopping...", self.get_info())
    }
}

fn main() {
    // Run the example
    client_code();
}
